### PR TITLE
Fix matched pan zoom

### PR DESCRIPTION
**Pull request by bmorris3**
### Description

This PR fixes matched pan/zoom in Imviz. [🐱](https://jira.stsci.edu/browse/JDAT-4088)

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 1]
```diff
@@ -351,36 +354,54 @@ def add_orientation(self, rotation_angle=None, east_left=None, label=None,
...
-        for data_label in self.orientation.choices:
-            self._add_data_to_all_viewers(data_label)
+        self._send_wcs_layers_to_all_viewers(viewers_to_update=[msg._viewer_id])
+
+    @observe('viewer_items')
+    def _send_wcs_layers_to_all_viewers(self, *args, **kwargs):
+        if not hasattr(self, 'viewer'):
+            return
+
+        wcs_only_layers = self.app.get_viewer_by_id('imviz-0').state.wcs_only_layers
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]: should this use the `default_image_viewer` thing instead of hardcoding `imviz-0`?

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 2]
```diff
@@ -351,36 +354,54 @@ def add_orientation(self, rotation_angle=None, east_left=None, label=None,
...
 
-        if hasattr(self, 'orientation') and len(self.orientation.choices):
-            self.viewer.selected = msg._viewer_id
-            self.orientation.selected = self.orientation.choices[0]
+        for viewer_ref in viewers_to_update:
+            for wcs_layer in wcs_only_layers:
+                if wcs_layer not in self.app.get_viewer_by_id(viewer_ref).state.wcs_only_layers:
+                    self.app.add_data_to_viewer(viewer_ref, wcs_layer)
+
+            self.viewer.selected = viewer_ref
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]: this will actually cycle the viewer select in the orientation plugin, right?  Is that for the sake of having the choices update in `self.orientation`?  Why do we need to do that now for all `viewers_to_update`?

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 3]
```diff
@@ -351,36 +354,54 @@ def add_orientation(self, rotation_angle=None, east_left=None, label=None,
...
+                self.orientation.selected = base_wcs_layer_label
 
     def _on_data_add_to_viewer(self, msg):
-        if (
-            msg._viewer_reference != 'imviz-0' and
-            self.app.get_viewer_by_id(msg._viewer_reference).state.reference_data is None
-        ):
-            self.viewer.selected = msg._viewer_reference
-            self.orientation.selected = base_wcs_layer_label
+        if msg._viewer_id != 'imviz-0':
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]: Again, should we use `default_image_viewer` to avoid the hardcoded `imviz-0`?

But also, would this ever be needed if `imviz-0` is emptied by the user first?

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 4]
```diff
@@ -406,10 +427,15 @@ def _on_refdata_change(self, msg={}):
             elif not len(viewer.data()):
                 self.link_type_selected = link_type_msg_to_trait['pixels']
 
+            if msg._data.label not in self.orientation.choices:
+                return
+
+            self.orientation.selected = msg._data.label
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]: why access `._data` instead of `.data` here?

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: Yeah, something like `self.app._jdaviz_helper.default_viewer.state.wcs_only_layers` would work, I think.

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 5]
```diff
@@ -392,8 +413,8 @@ def _change_reference_data(self, *args, **kwargs):
                 self.orientation.selected, viewer_id=self.viewer.selected
             )
 
-    def _on_refdata_change(self, msg={}):
-        self.orientation.only_wcs_layers = msg.data.meta.get('_WCS_ONLY', False)
+    def _on_refdata_change(self, msg):
+        self.orientation.set_wcs_only_filter(wcs_only=msg.data.meta.get('_WCS_ONLY', False))
         if hasattr(self, 'viewer'):
             ref_data = self.ref_data
             viewer = self.app.get_viewer(self.viewer.selected)
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]: Using`self.viewer.selected_obj` here will be slightly more efficient since it will cache

File: jdaviz/configs/imviz/plugins/orientation/orientation.py [hunk 6]
```diff
@@ -406,10 +427,15 @@ def _on_refdata_change(self, msg={}):
...
+            if msg._data.label not in self.orientation.choices:
+                return
+
+            self.orientation.selected = msg._data.label
+
             # we never want to highlight subsets of pixels within WCS-only layers,
             # so if this layer is an ImageSubsetLayerState on a WCS-only layer,
             # ensure that it is never visible:
-            for layer in viewer.state.layers:
+            for layer in self.viewer.selected_obj.state.layers:
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]: hmmm, I think this should be the same as `viewer` defined above (with or without the suggestion to use `viewer.selected_obj` there as well).

File: jdaviz/core/template_mixin.py [hunk 7]
```diff
@@ -1463,6 +1462,22 @@ def _sort_by_icon(items_dict):
 
         self._apply_default_selection()
 
+    def set_wcs_only_filter(self, wcs_only):
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py]: I'm still not convinced this belongs here vs in orientation plugin, but can see arguments both ways.  `set` might be confusing though... how about `toggle` or separate `add`/`remove` methods and then do the check externally?

Either way, this code is going to be a bit difficult to parse in the future, so could use a docstring for future us.

File: jdaviz/app.py [hunk 8]
```diff
@@ -2254,7 +2255,7 @@ def _create_viewer_item(self, viewer, vid=None, name=None, reference=None):
             'config': self.config,  # give viewer access to app config/layout
             'data_open': False,
             'collapse': True,
-            'reference': reference,
+            'reference': reference or name or vid,
```
Comment (kecnry) [inline jdaviz/app.py:2258]: why was this needed (specifically for this PR - I agree we should try to merge these eventually, I just worry this might add more confusion since now we'll have "reference" in the viewer item that isn't necessarily the same as "reference" on the viewer itself).

Comment (bmorris3) [reply jdaviz/app.py:2258]: This is needed because when a new image viewer is created, there are hooks in place to set `vid` and `name`, but not `reference` (https://github.com/spacetelescope/jdaviz/blob/6372879516cdb96081e90b4b5ced404dc66448fe/jdaviz/configs/imviz/helper.py#L54C28-L56), but the `reference` is the version used in some places.

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 1 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: Works for me.
```suggestion
        wcs_only_layers = self.app._jdaviz_helper.default_viewer.state.wcs_only_layers
```

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 3 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: Since the user isn't allowed to delete `imviz-0`, this probably isn't necessary, I wrote this conditional as part of logic that was unnecessary, so I can remove it now.

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 5 above)
Comment (bmorris3) [inline jdaviz/configs/imviz/plugins/orientation/orientation.py]:
```suggestion
            viewer = self.viewer.selected_obj
```

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 4 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: Only because I referred to the attrs in `__init__` without remembering the properties. 🤦‍♂️
```suggestion
            if msg.data.label not in self.orientation.choices:
                return

            self.orientation.selected = msg.data.label

```

File: jdaviz/core/template_mixin.py (hunk 7 above)
Comment (bmorris3) [reply jdaviz/core/template_mixin.py]: I'll do `:s/set/update/`, and add a docstring 👍🏻

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 2 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: We need that because we call `self.orientation.update_wcs_only_filter` below, which is particular to each viewer.

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: Does this work still?  That might expose the public API, so you might need `._obj` in there now (if you have https://github.com/spacetelescope/jdaviz/pull/2563, and otherwise will need it on rebase).

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 2 above)
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: but `self.orientation` won't remember per-viewer filters, so setting for anything but the last iteration of the for-loop will just be wasted effort... I think

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: The one time that this is necessary is when a new viewer is created and the WCS-only reference data have been added to the viewer but not yet selected as reference data. We first need to change the filter to "see" the orientation options, then we need to select the default orientation – I do this a few lines below (https://github.com/bmorris3/jdaviz/pull/29/files/e2f36a547e94cb0d38e45b807cb7dfba38c85122#diff-022bcba39f3fd68183229b5433146c2a21ac800db2778790152edc9a6b1b01caR398).

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 1 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: that's right – I pushed a correction.

File: jdaviz/configs/imviz/plugins/orientation/orientation.py (hunk 2 above)
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: ok, and in that case `viewers_to_update` is passed and only has one entry it seems.  I wonder if instead of having 

```
self.viewer.selected = viewer_ref
...
self.orientation.selected = base_wcs_layer_label
```

in the for-loop and run for all (passed) viewers, we should instead have 

```
if self.viewer.selected == viewer:
    ....
    self.orientation.selected = base_wcs_layer_label
```

Does that make sense?

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/orientation/orientation.py]: I understand what you're going for, but this method is _also_ used when new orientation layers are created. In that case, it adds the new orientation layer to each viewer, so that it can be selected as refdata in that viewer. So the function is written for the general case.

File: jdaviz/app.py [hunk 9]
```diff
@@ -2319,6 +2320,12 @@ def _on_new_viewer(self, msg, vid=None, name=None, add_layers_to_viewer=False):
         ref_data = self._jdaviz_helper.default_viewer._obj.state.reference_data
         new_viewer_item['reference_data_label'] = getattr(ref_data, 'label', None)
 
+        if (
+                getattr(ref_data, 'meta', {}).get('_WCS_ONLY', False) and
```
Comment (pllim) [inline jdaviz/app.py]: Is it more consistent to use `layer_is_wcs_only` function here that you added in https://github.com/spacetelescope/jdaviz/pull/2179 in Imviz helper, or is there a reason to not do that here?

File: jdaviz/configs/imviz/plugins/tools.py [hunk 10]
```diff
@@ -51,7 +51,10 @@ def on_click(self, data):
         if x is None or y is None:  # Out of bounds
             return
         x, y, _, _ = self.viewer._get_real_xy(image, x, y)
-        self.viewer.center_on((x, y))
+        if image.coords is not None:
+            self.viewer.center_on(image.coords.pixel_to_world(x, y))
```
Comment (pllim) [inline jdaviz/configs/imviz/plugins/tools.py]: Why this logic? It kinda defeats the purpose of `_get_real_xy` above.

File: jdaviz/configs/imviz/tests/test_tools.py
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_tools.py:1]: Why did the results change?

File: jdaviz/configs/imviz/wcs_utils.py [hunk 11]
```diff
@@ -440,12 +440,10 @@ def _prepare_rotated_nddata(real_image_shape, wcs, rotation_angle, refdata_shape
 
     # create a fake NDData (we use arange so data boundaries show up in Imviz
     # if it ever is accidentally exposed) with the rotated GWCS:
-    sequential_data = np.arange(
-        np.prod(refdata_shape), dtype=np.int8
-    ).reshape(refdata_shape)
+    placeholder_data = np.nan * np.ones(refdata_shape)
```
Comment (pllim) [inline jdaviz/configs/imviz/wcs_utils.py:443]: What is the benefit of changing sequential to NaN? Kinda make it hard to debug when the fake layer accidentally shows up.

File: jdaviz/core/astrowidgets_api.py [hunk 12]
```diff
@@ -178,8 +178,11 @@ def zoom_level(self):
             raise ValueError('Viewer is still loading, try again later')
 
         if hasattr(self, '_get_real_xy'):
-            i_top = get_top_layer_index(self)
-            image = self.layers[i_top].layer
+            if self.state.reference_data is not None:
+                image, i_ref = get_reference_image_data(self.jdaviz_app, self.reference)
```
Comment (pllim) [inline jdaviz/core/astrowidgets_api.py:186]: I don't understand this change. Zoom should be based on the visible layer, not the hidden reference data.

File: jdaviz/core/template_mixin.py [hunk 13]
```diff
@@ -1463,6 +1462,35 @@ def _sort_by_icon(items_dict):
...
+        the layer choices by adjusting the layer filters on this
+        LayerSelect instance.
+
+        Parameters
+        ----------
+        wcs_only : bool
+            `True` will filter only the WCS-only layers, `False` will
+            give the non-WCS-only layers.
+        """
+        def is_wcs_only(data):
```
Comment (pllim) [inline jdaviz/core/template_mixin.py]: I feel like this re-implemented `layer_is_wcs_only` that you also implemented in Imviz helper.

File: jdaviz/core/template_mixin.py [hunk 14]
```diff
@@ -1463,6 +1462,35 @@ def _sort_by_icon(items_dict):
...
+        filter_names = [getattr(filt, '__name__', '') for filt in self.filters]
+
+        if not wcs_only and 'is_wcs_only' in filter_names:
+            self.filters.remove(*[filt for filt in self.filters
+                                  if getattr(filt, '__name__', '') == 'is_wcs_only'])
+        elif wcs_only and 'is_wcs_only' not in filter_names:
+            self.add_filter(is_wcs_only)
+
+    @property
+    def only_wcs_layers(self):
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:1478]: How is this different from `get_wcs_only_layer_labels`? Please clarify in a docstring.

File: jdaviz/core/tools.py [hunk 15]
```diff
@@ -102,45 +101,37 @@ def on_limits_change(self, *args):
...
+            else:
+                to_fov_sky = None
+
+            if to_fov_sky is not None and orig_fov_sky is not None:
+                old_level = viewer.zoom_level
+                viewer.zoom_level = old_level * float(to_fov_sky / orig_fov_sky)
+                viewer.center_on(sky_cen)
+
+            else:
+                with delay_callback(viewer.state, *self.match_keys):
```
Comment (pllim) [inline jdaviz/core/tools.py:117]: Some comments to explain the intention of this logic block would ease future debugging, in case this math does not work for all use cases.

**Review by pllim — COMMENTED** [review]
Technical review as follows. Thanks!

File: jdaviz/configs/imviz/tests/test_tools.py
Comment (bmorris3) [reply jdaviz/configs/imviz/tests/test_tools.py:1]: These test results have changed because the matched pan/zoom approach for data with WCS is slightly different than before, producing limits different by one pixel. The main differences come from trying to match zoom approximately even when viewers have different reference data.

File: jdaviz/configs/imviz/wcs_utils.py (hunk 11 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/wcs_utils.py:443]: This was discussed in an earlier PR that we merged, but then I botched a rebase, so I'm fixing that mistake in this PR. Since this was approved before, I'm just reinserting it.

There are two main benefits if the "WCS-only reference data" _have no data_: 
1) when a WCS-only layer is created, it is visible for a few milliseconds, before the visibility is set to `False`. If the data are `nan`, they are never rendered, and the viewer does less crazy flashing during rotation. 
2) If you somehow do math on the WCS-only layer's pixels, the result will never appear to be legitimate. This should not be easy to do, given the checks we have in place, but switching to `nan` makes it much easier to debug if we accidentally do operations on the WCS-only data.

File: jdaviz/core/template_mixin.py (hunk 13 above)
Comment (bmorris3) [reply jdaviz/core/template_mixin.py]: This needs to be a separate function that you can turn on and off as a `filter` for `LayerSelect`.

File: jdaviz/core/astrowidgets_api.py (hunk 12 above)
Comment (bmorris3) [reply jdaviz/core/astrowidgets_api.py:186]: > Zoom should be based on the visible layer

Why? If they're linked, is this distinction meaningful?

File: jdaviz/configs/imviz/plugins/tools.py (hunk 10 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/tools.py]: The use of `_get_real_xy` above ensures that the behavior for data with WCS (in the `else`) and without WCS (in the `if`) is consistent. `image` here is [the first visible layer](https://github.com/bmorris3/jdaviz/blob/131a4123e6e03fbd505dc9992a947c5f85510d0f/jdaviz/configs/imviz/plugins/tools.py#L48), which is not the reference data layer when linked by WCS. So this call tries to be explicit.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py]: If you want to switch the call to using sky coordinates, you don't need `_get_real_xy`. You can theoretically pass `(x, y)` directly to reference WCS because input `(x, y)` is always w.r.t. reference data.

File: jdaviz/configs/imviz/tests/test_tools.py
Comment (pllim) [reply jdaviz/configs/imviz/tests/test_tools.py:1]: I have to think about this:

When you give `'domain': {'x': 1, 'y': 1}`, that is where you are telling the reference data to go. In this case, `v` has both images loaded, so second image is on top. Second image is offset by +1 pix in the WCS. So if the adjusted limits reflects second image limits, then I think this change makes sense.

But then on L63-L62, `v2` should only have the first image, and both `v` and `v2` give the same limits. Then the assumption above breaks down. Let me run this example interactively and see what is actually going on.

File: jdaviz/core/astrowidgets_api.py (hunk 12 above)
Comment (pllim) [reply jdaviz/core/astrowidgets_api.py:186]: Because I want the zoom level of the image under view, not zoom level of the reference data.

Comment (pllim) [reply jdaviz/core/astrowidgets_api.py:186]: But the result in example notebook seems correct (when I installed this branch rebased on top of main). Let me think more.

Comment (pllim) [reply jdaviz/core/astrowidgets_api.py:186]: Ah, no, it is completely wrong. I forgot to link by WCS when I commented 8 mins ago. When I link by WCS, the zoom_level is wrong. This change needs to be reverted.

Comment (pllim) [reply jdaviz/core/astrowidgets_api.py:186]: To give you a sense of what I expect, in the ImvizDitheredExample notebook, a "fit" zoom would be roughly 0.2. A `imviz.default_viewer.zoom_level` of 1 would be like this:

Screenshot 2024-01-12 125400

It should not matter if I am linking by WCS or not.

File: jdaviz/configs/imviz/tests/test_tools.py
Comment (pllim) [reply jdaviz/configs/imviz/tests/test_tools.py:1]: I still don't get this change. This test case is linked by pixel. Even though both images have WCS, it should not have mattered. No orientation layer was created.

With this PR, when I run this test case interactively, this centered the image at `x=2.5 y=1.5`:

```python
t_normpan = v.toolbar.tools['jdaviz:imagepanzoom']
t_normpan.activate()
t_normpan.on_click({'event': 'click', 'domain': {'x': 1, 'y': 1}})
```

And if I have the pan/zoom tool active, where a single click would center to the pixel under cursor, that pixel is not actually the center when "pan to click location" is done.

I do not think this change in test behavior is correct.

File: jdaviz/configs/imviz/plugins/tools.py (hunk 10 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/tools.py]: So do you want to see line 53 moved inside the conditional?

File: jdaviz/configs/imviz/tests/test_tools.py
Comment (pllim) [reply jdaviz/configs/imviz/tests/test_tools.py:1]: On `main`, the same test case does not have this x offset. Pixel under cursor is centered though I suspect there might be a 0.5 offset in y but hard to tell without built in grids on the viewer.

File: jdaviz/configs/imviz/wcs_utils.py (hunk 11 above)
Comment (pllim) [reply jdaviz/configs/imviz/wcs_utils.py:443]: Yes, I think the benefit not having the weird flashing is good. Thanks for the clarification!

File: jdaviz/configs/imviz/plugins/tools.py (hunk 10 above)
Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py]: If this line makes the final cut, I think you can change the call to something like `reference_data.coords.pixel_to_world(x, y)` without ever calling `viewer._get_real_xy(image, ...)`.

But I am not sure what you need to do to fix https://github.com/bmorris3/jdaviz/pull/29#discussion_r1450789355

**Comment by bmorris3** [comment]
@pllim [5c3d48b](https://github.com/bmorris3/jdaviz/pull/29/commits/5c3d48b46a614308a86196b00ff130586125a59c) reverts the change to the `center_on` call, which allows me to revert the expected test results successfully.

File: jdaviz/configs/imviz/plugins/tools.py (hunk 10 above)
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/tools.py]: This line is already out of the final cut 👍🏻

**Comment by pllim** [comment]
It is unfortunate the remote data job is failing for a different reason, so we cannot completely convince ourselves via CI until after rebase.

**Review by pllim — COMMENTED** [review]
Also, any chance we can turn Cami's use case that started all this into a test?

I already forgot her workflow. 😆

File: jdaviz/app.py (hunk 9 above)
Comment (bmorris3) [reply jdaviz/app.py]: I had avoided doing that only because of circular imports, originally. But I've pushed [this commit](https://github.com/bmorris3/jdaviz/pull/29/commits/ea13ceacc60678a9fc67ca09d9880d452d268eaf) which imports the function within the method and dodges the circularity.

File: jdaviz/core/astrowidgets_api.py [hunk 16]
```diff
@@ -178,8 +182,11 @@ def zoom_level(self):
...
```
Comment (pllim) [inline jdaviz/core/astrowidgets_api.py:186]: `i_ref` is not used.
```suggestion
                image, _ = get_reference_image_data(self.jdaviz_app, self.reference)
```

File: jdaviz/core/astrowidgets_api.py [hunk 17]
```diff
@@ -210,8 +217,11 @@ def zoom_level(self, val):
 
         new_dx = self.shape[1] * 0.5 / val
         if hasattr(self, '_get_real_xy'):
-            i_top = get_top_layer_index(self)
-            image = self.layers[i_top].layer
+            if self.state.reference_data is not None:
+                image, i_ref = get_reference_image_data(self.jdaviz_app, self.reference)
```
Comment (pllim) [inline jdaviz/core/astrowidgets_api.py:221]:
```suggestion
                image, _ = get_reference_image_data(self.jdaviz_app, self.reference)
```

**Review by pllim — COMMENTED** [review]
When you run the ImvizDitheredExample notebook (on this branch without rebasing on top of main), linked by WCS, 2 viewers, zoomed in a little -- Do you see the second viewer flicker when you linked pan/zoom the first one? I see it flicker a few times before settling.

Also, once I switch the tool from linked to normal pan/zoom (or box zoom), I can no longer switch back, as in when I right-click on the tool, there is no linked option anymore.

Screenshot 2024-01-12 170528

Neither of these are a problem on `main`.

**Comment by bmorris3** [comment]
@pllim I'm hopeful [3c5fb47](https://github.com/bmorris3/jdaviz/pull/29/commits/3c5fb47fb9820b831263c87afc61c4d867b62816) fixed up the data menu problems. Seems to work for me. 🤞🏻

File: jdaviz/core/astrowidgets_api.py [hunk 18]
```diff
@@ -166,7 +166,11 @@ def offset_by(self, dx, dy):
 
     @property
     def zoom_level(self):
-        """Zoom level:
+        """
+        The zoom level for an image viewer (not linked by WCS).
+
+        Warning: when a viewer is linked by WCS, the result corresponds
```
Comment (pllim) [inline jdaviz/core/astrowidgets_api.py]: Make it more obvious.
```suggestion
        .. warning:: when a viewer is linked by WCS, the result corresponds
```

File: jdaviz/core/astrowidgets_api.py [hunk 19]
```diff
@@ -178,8 +182,11 @@ def zoom_level(self):
             raise ValueError('Viewer is still loading, try again later')
 
         if hasattr(self, '_get_real_xy'):
-            i_top = get_top_layer_index(self)
-            image = self.layers[i_top].layer
+            if self.state.reference_data is not None:
+                image, i_ref = get_reference_image_data(self.jdaviz_app, self.reference)
+            else:
```
Comment (pllim) [inline jdaviz/core/astrowidgets_api.py:187]: Is this `else` ever used? I don't think we would encounter a viewer without reference data but has a top layer displayed.

File: jdaviz/core/tools.py [hunk 20]
```diff
@@ -102,45 +101,37 @@ def on_limits_change(self, *args):
...
+
+            if hasattr(viewer, '_get_fov'):
+                to_fov_sky = viewer._get_fov(wcs=matched_refdata.coords)
+            else:
+                to_fov_sky = None
+
+            if to_fov_sky is not None and orig_fov_sky is not None:
+                old_level = viewer.zoom_level
+                viewer.zoom_level = old_level * float(to_fov_sky / orig_fov_sky)
+                viewer.center_on(sky_cen)
```
Comment (pllim) [inline jdaviz/core/tools.py:114]: This is the part I was referring to when I say you can roll your own calculations if we ever decide that Astrowidgets API cannot make the results here reasonable. cc @kecnry

**Comment by pllim** [comment]
@bmorris3 , alas, does not work for me. Maybe another dev can check if it is just me or it is really still a problem.

Screenshot 2024-01-16 184618

**Comment by pllim** [comment]
Could be just me because now even on the new machine, the weird data menu behavior that somehow only plagues me in the old machine is appearing again. 😭 

Screenshot 2024-01-16 185017

**Comment by kecnry** [comment]
@pllim - that first screenshot looks like it was fixed by https://github.com/spacetelescope/jdaviz/pull/2605 (the main image rotation PR probably just needs a rebase to get that in).

**Review by pllim — APPROVED** [review]
Why don't we just merge then, so rebase can happen? Thanks!
